routing.go

```diff
@@ -121,111 +121,200 @@ func (dht *IpfsDHT) GetValue(ctx context.Context, key string, opts ...ropts.Opti
+	responses, err := dht.SearchValue(ctx, key, opts...)
 	if err != nil {
 		return nil, err
 	}
+	var best []byte
 
-	recs := make([][]byte, 0, len(vals))
-	for _, v := range vals {
-		if v.Val != nil {
-			recs = append(recs, v.Val)
-		}
+	for r := range responses {
+		best = r
 	}
-	if len(recs) == 0 {
+
+	if ctx.Err() != nil {
+		return best, ctx.Err()
+	}
+
+	if best == nil {
 		return nil, routing.ErrNotFound
 	}
+	log.Debugf("GetValue %v %v", key, best)
+	return best, nil
+}
 
-	i, err := dht.Validator.Select(key, recs)
-	if err != nil {
+func (dht *IpfsDHT) SearchValue(ctx context.Context, key string, opts ...ropts.Option) (<-chan []byte, error) {
+	var cfg ropts.Options
+	if err := cfg.Apply(opts...); err != nil {
 		return nil, err
 	}
 
-	best := recs[i]
-	log.Debugf("GetValue %v %v", key, best)
-	if best == nil {
-		log.Errorf("GetValue yielded correct record with nil value.")
-		return nil, routing.ErrNotFound
+	responsesNeeded := 0
+	if !cfg.Offline {
+		responsesNeeded = getQuorum(&cfg, -1)
 	}
 
-	fixupRec := record.MakePutRecord(key, best)
-	for _, v := range vals {
-		// if someone sent us a different 'less-valid' record, lets correct them
-		if !bytes.Equal(v.Val, best) {
-			go func(v RecvdVal) {
-				if v.From == dht.self {
-					err := dht.putLocal(key, fixupRec)
-					if err != nil {
-						log.Error("Error correcting local dht entry:", err)
-					}
+	valCh, err := dht.getValues(ctx, key, responsesNeeded)
+	if err != nil {
+		return nil, err
+	}
+
+	out := make(chan []byte)
+	go func() {
+		defer close(out)
+
+		maxVals := responsesNeeded
+		if maxVals < 0 {
+			maxVals = defaultQuorum * 4 // we want some upper bound on how
+			// much correctional entries we will send
+		}
+		vals := make([]RecvdVal, 0, maxVals)
+		best := -1
+
+		defer func() {
+			if len(vals) <= 1 || best < 0 {
+				return
+			}
+			fixupRec := record.MakePutRecord(key, vals[best].Val)
+			for _, v := range vals {
+				// if someone sent us a different 'less-valid' record, lets correct them
+				if !bytes.Equal(v.Val, vals[best].Val) {
+					go func(v RecvdVal) {
+						if v.From == dht.self {
+							err := dht.putLocal(key, fixupRec)
+							if err != nil {
+								log.Error("Error correcting local dht entry:", err)
+							}
+							return
+						}
+						ctx, cancel := context.WithTimeout(dht.Context(), time.Second*30)
+						defer cancel()
+						err := dht.putValueToPeer(ctx, v.From, key, fixupRec)
+						if err != nil {
+							log.Debug("Error correcting DHT entry: ", err)
+						}
+					}(v)
+				}
+			}
+		}()
+
+		for {
+			select {
+			case v, ok := <-valCh:
+				if !ok {
 					return
 				}
-				ctx, cancel := context.WithTimeout(dht.Context(), time.Second*30)
-				defer cancel()
-				err := dht.putValueToPeer(ctx, v.From, key, fixupRec)
-				if err != nil {
-					log.Debug("Error correcting DHT entry: ", err)
+
+				i := len(vals)
+				if len(vals) < maxVals {
+					vals = append(vals, v)
+				} else {
+					i = (best + 1) % maxVals
```

Comment:
`i` could still be out of bounds. It may be better to simply put the "best" value in a separate variable and then keep an array of peers to correct.